### ISSUE TITLE
Websocket cleanup

### ISSUE THREAD

**Issue by BradForsythe** [dougdellolio/gdax-csharp#114]
Doug, just looking at using your websocket interface instead of my own since it's a lot cleaner and nicely integrated.   A few comments/notes for consideration...
1.  Your public control interface is Start/Stop.  Start will authenticate and request subscriptions.  Great!  Stop looks like the socket closes but technically speaking we are still subscribed.  You may want to consider an unsubscribe then close.
2.  Upon closing the host app after performing a stop, feed data may still work its way in and you throw a socket error exception.  It would be nice to put in a handler or provide a public dispose function on the socket.   Something like gsharpClient.Dispose which ultimately calls the Web4Net socket dispose since sockets are not part of the managed garbage collector.   This solves the exception error.   Maybe unsubscribing to the data also solves the problem?
3.  Would also be nice to have a public handler for OnWebSocket_Close, Open and Error.  This helps to monitor status of the socket.

Most other issues such as feed loss detection can be done outside the client.  Overall looks good though.

It's likely I may have missed something if I am off base.

**Comment by Sotam** [comment]
You are both right about the first and second one. I should've looked more careful when implementing the web socket feed. Currently it also not possible to subscribe to new channels while you've already started the web socket feed. Will try take a look in the next few days :)

And about the last one, also something nice to have. Not really difficult to add, @dougdellolio please let me know you think of it and if you mind that I add/fix these points.

**Comment by dougdellolio** [comment]
@BradForsythe thanks for the feedback. i am in the process of doing some refactoring separate of the points you have provided and will have a PR shortly.

@Sotam if you're up for it  please do!

**Comment by BradForsythe** [comment]
Awesome.   While you are looking at the feed, you may want to consider adding to your OrderTypes  Order life cycle for standard orders go through received, open, match, then done.   You have some of the handlers for these but not the types.   Really, just open and done are about all you need on an order that was accepted by the matching engine.  The rest are just to be complete.

**Comment by BradForsythe** [comment]
Also....  :) ....  On your Ticker channel response (websocket) you are returning the GDAX string typed product ID  i.e.  :"BTC-USD" and not your product type.   No biggie, but it would be nice to have an enum extension to reverse lookup the product ID given the EnumMember attribute.

Like:   ProductType pt = EnumMemberDictionary<ProductType>.GetValue(e.LastOrder.ProductId);

First time it is called, the dictionary is automatically created.  I call it at the constructor of my socket service (about 24msec)  Then during runtime handling Ticker it runs quick (<1ms).
 
A simple dictionary to do the reverse look up:

    public static class EnumMemberDictionary<T> where T : enumDict
    {
        private static readonly Dictionary<string, T> enumDict =
            new Dictionary<string, T>();
        static EnumMemberDictionary()
        {    
            // Could do this with a LINQ query, admittedly...
            foreach (var field in typeof(T).GetFields
                     (BindingFlags.Public | BindingFlags.Static))
            {
                T value = (T)field.GetValue(null);
                foreach (var description in (EnumMemberAttribute[])
                   field.GetCustomAttributes(typeof(EnumMemberAttribute), false))
                {
                    enumDict [description.Value] = value;
                }
            }
        }
        public static T GetValue(string description)
        {
            T ret;
            if (enumDict.TryGetValue(description, out ret))
            {
                return ret;
            }
            return default(T);
        }
    }

**Comment by Sotam** [comment]
If I understand you correctly (@BradForsythe), such a dictionary is not necessary. Newtonsoft's JSON can easily convert (not sure if there is a difference in performance). So changing `string` to `ProductType` should be enough, will add it in the next commit 😄 

E.g.
https://github.com/dougdellolio/gdax-csharp/blob/d24408d88f27f933d43595864cc352d1b4600cf8/GDAXSharp/WebSocket/Models/Response/Heartbeat.cs#L9

**Comment by dougdellolio** [comment]
no worries @BradForsythe. If you feel up to contributing, please do as it may benefit others too otherwise I can get to some of these things soon

**Comment by BradForsythe** [comment]
No problem, I'd be happy to do.  I'll add in some handlers first.   That's pretty straightforward and then sort out the unsubscribing on socket close.

**Comment by BradForsythe** [comment]
After a bit of "Git pain/learning" (I am more experienced with SVN)  and  realizing each time I commit the changes are pushed with the PR even work in progress commits.   Maybe I'll read the instruction manual.  :)

I updated the WebSocket to include public handlers for the OnClose, OnOpen and OnError events.   Also I added an Unsubcribe on closing the socket.   Note:  the tests for OnClose are not correct, I put in an unnecessary check for state to be closed in the OnClosed event handler.  But I didn't want to change the test...because I am not sure if I was suppose to...

Future improvement would be to turn on and off subscriptions for select products and channels.   Useful if you want to keep you feeds to what you are mainly interested and then occasionally want to check for Trades for example.    This can be done by opening and closing the socket by the "app" but it is possible and valid to change subscriptions on the fly.
